### PR TITLE
Add IPC with shared memory instead of file creation.

### PR DESCRIPTION
**Pull request by ehartmann**

### CODE REVIEW

File: server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java [hunk 1]
```diff
@@ -42,75 +43,71 @@
...
+  private final RandomAccessFile sharedmemory;
+  /**
+   * The ByteBuffer will contains :
+   * <ul>
+   *   <li>First byte will contains 0x00 until stop command is issued = 0xFF</li>
+   *   <li>Then each 9 bytes will contains first the status of process (0x01 : READY) then a long (8 bytes) with the latest ping</li>
+   * </ul>
+   */
+  final MappedByteBuffer mappedByteBuffer;
+  private static final Integer MAX_SHARED_MEMORY = 1 + 9 * 10; // We this shared memory we can handle up to 10 processes (91 bytes)
```
Comment (simonbrandhof) [inline server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java]: typo: "we this shared memory" --> "with this shared memory" ?

File: server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java [hunk 2]
```diff
@@ -42,75 +43,71 @@
  */
 public class ProcessCommands {
 
-  private final File readyFile, stopFile;
+  private final RandomAccessFile sharedmemory;
```
Comment (simonbrandhof) [inline server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java]: This field can be moved as local variable

File: server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java (hunk 1 above)
Comment (simonbrandhof) [inline server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java]: Integer can be replaced by int

File: server/sonar-process/src/main/java/org/sonar/process/StopWatcher.java [hunk 3]
```diff
@@ -45,23 +45,19 @@ public StopWatcher(ProcessCommands commands, Stoppable stoppable) {
...
   @Override
   public void run() {
-    commands.prepare();
-    try {
-      while (watching) {
-        if (commands.askedForStop()) {
-          LoggerFactory.getLogger(getClass()).info("Stopping process");
-          stoppable.stopAsync();
+    while (watching) {
+      commands.ping();
```
Comment (simonbrandhof) [inline server/sonar-process/src/main/java/org/sonar/process/StopWatcher.java]: why does it ping when asking for stop ?

File: sonar-application/src/main/java/org/sonar/application/App.java [hunk 4]
```diff
@@ -56,7 +56,7 @@ public void start(Props props) {
     if (props.valueAsBoolean(ProcessConstants.ENABLE_STOP_COMMAND, false)) {
       // stop application when file <temp>/app.stop is created
       File tempDir = props.nonNullValueAsFile(ProcessConstants.PATH_TEMP);
-      ProcessCommands commands = new ProcessCommands(tempDir, "app");
+      ProcessCommands commands = new ProcessCommands(tempDir, 1);
```
Comment (simonbrandhof) [inline sonar-application/src/main/java/org/sonar/application/App.java]: Index is an internal information used only by process monitor. I think indices should be hidden from caller and should automatically generated and incremented in Monitor#start().

File: server/sonar-process/src/main/java/org/sonar/process/StopWatcher.java (hunk 3 above)
Comment (ehartmann) [inline server/sonar-process/src/main/java/org/sonar/process/StopWatcher.java]: If my understanding is correct, it's the thread that monitor if there is an askedForStop every 500ms. I though it's the best way to both monitor the flag and update the field containing the last ping command. I think that we do not need another thread just for pinging.

File: server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java (hunk 2 above)
Comment (ehartmann) [inline server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java]: Exact, I'll update the code.

File: server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java (hunk 1 above)
Comment (ehartmann) [inline server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java]: Exact, I'll update the code.

File: server/sonar-process/src/main/java/org/sonar/process/StopWatcher.java (hunk 3 above)
Comment (simonbrandhof) [inline server/sonar-process/src/main/java/org/sonar/process/StopWatcher.java]: Two options:
* rename the class as it does not do only stop watching but also pinging
* remove ping() as it's still not used for autokill
I'm for option 2.

File: server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java [hunk 5]
```diff
@@ -42,75 +45,99 @@
...
+      return -1;
+    }
   }
 
   /**
    * To be executed by monitor process to ask for child process termination
    */
   public void askForStop() {
-    createFile(stopFile);
+    mappedByteBuffer.put(offset() + 1, STOP);
```
Comment (simonbrandhof) [inline server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java]: you should document why you add 1

File: server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java [hunk 6]
```diff
@@ -42,75 +45,99 @@
...
   }
 
   public boolean askedForStop() {
-    return stopFile.exists();
+    return mappedByteBuffer.get(offset() + 1) == STOP;
```
Comment (simonbrandhof) [inline server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java]: you should document why you add 1

File: server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java (hunk 5 above)
Comment (ehartmann) [inline server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java]: Done

File: server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java (hunk 6 above)
Comment (ehartmann) [inline server/sonar-process/src/main/java/org/sonar/process/ProcessCommands.java]: Done

**Pull request merged**